src/DeploymentFile.ts

```diff
@@ -0,0 +1,135 @@
+    private _jsonParseResult: Json.ParseResult;
+
+    // The JSON node for the top-level JSON object (if the JSON is not empty or malformed)
+    private _topLevelValue: Json.ObjectValue | undefined;
+
+    private _schema: CachedValue<Json.StringValue | undefined> = new CachedValue<Json.StringValue | undefined>();
+
+    /**
+     * Constructor
+     *
+     * @param _documentText The string text of the document.
+     * @param _documentId A unique identifier for this document. Usually this will be a URI to the document.
+     */
+    constructor(private _documentText: string, private _documentId: string) {
+        nonNullOrEmptyValue(_documentId, "_documentId");
+
+        this._jsonParseResult = Json.parse(_documentText);
+        this._topLevelValue = Json.asObjectValue(this._jsonParseResult.value);
+    }
+
+    /**
+     * Get the document text as a string.
+     */
+    public get documentText(): string {
+        return this._documentText;
+    }
+
+    /**
+     * The unique identifier for this deployment template. Usually this will be a URI to the document.
+     */
+    public get documentId(): string {
+        return this._documentId;
+    }
+
+    // Parse result for the template JSON document as a whole
+    public get jsonParseResult(): Json.ParseResult {
+        return this._jsonParseResult;
+    }
+
+    // The JSON node for the top-level JSON object (if the JSON is not empty or malformed)
+    public get topLevelValue(): Json.ObjectValue | undefined {
+        return this._topLevelValue;
+    }
+
+    public get schemaUri(): string | undefined {
+        const schema = this.schemaValue;
+        return schema ? schema.unquotedValue : undefined;
+    }
+
+    public get schemaValue(): Json.StringValue | undefined {
+        return this._schema.getOrCacheValue(() => {
+            const value: Json.ObjectValue | undefined = Json.asObjectValue(this._jsonParseResult.value);
+            if (value) {
+                const schema: Json.StringValue | undefined = Json.asStringValue(value.getPropertyValue("$schema"));
+                if (schema) {
+                    return schema;
+                }
+            }
+
+            return undefined;
+        });
+    }
+
+    public getMaxLineLength(): number {
+        let max = 0;
+        for (let len of this.jsonParseResult.lineLengths) {
+            if (len > max) {
+                max = len;
+            }
+        }
+
+        return max;
+    }
+
+    public getCommentCount(): number {
+        return this.jsonParseResult.commentCount;
+    }
+
+    /**
+     * Get the number of lines that are in the file.
+     */
+    public get lineCount(): number {
+        return this._jsonParseResult.lineLengths.length;
+    }
+
+    /**
+     * Get the maximum column index for the provided line. For the last line in the file,
+     * the maximum column index is equal to the line length. For every other line in the file,
+     * the maximum column index is less than the line length.
```

Comment:
>less than the line length [](start = 35, length = 25)

Because line length includes EOL? Also, what if there tabs on the line? One char/four columns?

Reply:
Not My Code (c).  But that appears to be correct.  Column really means "character index from start of line", so tabs are not an issue, they count as one character.